core/websock.js

```diff
@@ -173,49 +175,92 @@ export default class Websock {
         this._sQ = new Uint8Array(this._sQbufferSize);
     }
 
+    _getChannelStates(channelType) {
+        if (channelType === "WebSocket") {
+            return {
+                CONNECTING: WebSocket.CONNECTING,
+                OPEN: WebSocket.OPEN,
+                CLOSING: WebSocket.CLOSING,
+                CLOSED: WebSocket.CLOSED
+            };
+        }
+        if (channelType === "RTCDataChannel") {
+            // Constants pulled from RTCDataChannelState enum
+            // https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/readyState#RTCDataChannelState_enum
+            return {
+                CONNECTING: "connecting",
+                OPEN: "open",
+                CLOSING: "closing",
+                CLOSED: "closed"
+            };
+        }
+        throw new Error(`channelType: ${channelType} not recognized`);
+    }
+
     init() {
         this._allocate_buffers();
         this._rQi = 0;
-        this._websocket = null;
+        this._webChannel = null;
     }
 
-    open(uri, protocols) {
+    open({ uri, protocols, webChannel, channelType }) {
         this.init();
 
-        this._websocket = new WebSocket(uri, protocols);
-        this._websocket.binaryType = 'arraybuffer';
+        if (uri) {
+            this._webChannel = new WebSocket(uri, protocols);
+            this._channelType = "WebSocket";
+            this._channelStates = this._getChannelStates("WebSocket");
+        } else if (webChannel && channelType) {
+            this._webChannel = webChannel;
+            this._channelType = channelType;
+            this._channelStates = this._getChannelStates(channelType);
+        } else {
+            throw new Error(
+                `Expected to receive one of uri and optional protocols or webChannel and channelType`
+            );
+        }
 
-        this._websocket.onmessage = this._recv_message.bind(this);
-        this._websocket.onopen = () => {
-            Log.Debug('>> WebSock.onopen');
-            if (this._websocket.protocol) {
-                Log.Info("Server choose sub-protocol: " + this._websocket.protocol);
+        const onOpen = () => {
+            Log.Debug(`>> ${this._channelType}.onopen`);
+            if (this._webChannel.protocol) {
+                Log.Info("Server choose sub-protocol: " + this._webChannel.protocol);
             }
 
             this._eventHandlers.open();
-            Log.Debug("<< WebSock.onopen");
+            Log.Debug(`<< ${this._channelType}.onopen`);
         };
-        this._websocket.onclose = (e) => {
-            Log.Debug(">> WebSock.onclose");
+
+        this._webChannel.binaryType = "arraybuffer";
+
+        this._webChannel.onmessage = this._recv_message.bind(this);
```

Comment:
Since we're now given an existing object I think it's time we start being more well behaved and use `addEventListener()` instead.

Reply:
this is the one feedback I didn't do, wasn't sure exactly what you meant.

Reply:
instead of using `onmessage` you should be able to do `addEventListener('message', this._recv_message.bind(this));`

Reply:
Both WebSocket and RTCDataChannel expose `onmessage` to attach a function handler to be called, shouldn't we stick to using the provided methods?

Reply:
From what I've heard it seems the `onevent` functions are becoming deprecated in favor of `addEventListener()`. I haven't looked into this one in detail, but it seems the most future proof path.

If we would make this change we should remove those 4 functions from `rawChannelProps` as well.

Reply:
Such a refactor would touch multiple unrelated areas of the codebase and seems outside the scope of this PR, could we make an issue to track this instead?

Reply:
We don't need to go over the entire codebase right now, but I think it needs to be done for the WebSocket/RTCDataChannel object on this PR. The reason being that we no longer have exclusive control of this object. The caller might also want to set up event handlers, so we should use the mechanism that allows multiple handlers.

Reply:
I am not quite sure how to update the FakeWebSocket class to have an `addEventListener` and `dispatchEvent` method to test against.